File: data-structures/graph/typescript/adjacency-list-graph.ts
interface Graph {
	[key: string]: string[];
}

export class AdjacencyListGraph {
	private graph: Graph;

	constructor() {
		this.graph = {};
	}

	addVertex(vertex: string): void {
		this.graph[vertex] = [];
	}

	addEdge(vertex1: string, vertex2: string): void {
		if (!this.graph[vertex1] || !this.graph[vertex2]) {
			throw new Error('Vertex not found in the graph');
		}
		this.graph[vertex1].push(vertex2);
		this.graph[vertex2].push(vertex1);
	}

	removeVertex(vertex: string): void {
		if (!this.graph[vertex]) {
			throw new Error('Vertex not found in the graph');
		}

		const verticesWithEdgeToDelete = this.graph[vertex];

		verticesWithEdgeToDelete.forEach(vertexWithEdgeToDelete => {
			this.graph[vertexWithEdgeToDelete] = this.graph[
				vertexWithEdgeToDelete
			].filter(edge => edge !== vertex);
		});

		delete this.graph[vertex];
	}

	removeEdge(vertex1: string, vertex2: string): void {
		if (!this.graph[vertex1] || !this.graph[vertex2]) {
			throw new Error('Vertex not found in the graph');
		}

		this.graph[vertex1] = this.graph[vertex1].filter(edge => edge !== vertex2);

		this.graph[vertex2] = this.graph[vertex2].filter(edge => edge !== vertex1);
	}

	depthFirstSearchRecursive(initialVertex: string): Set<string> {
		if (!this.graph[initialVertex]) {
			throw new Error('Vertex not found in the graph');
		}

		const visited: Set<string> = new Set();
		const traverse = (vertex: string): void => {
			visited.add(vertex);
			this.graph[vertex].forEach(neighbor => {
				if (!visited.has(neighbor)) {
					return traverse(neighbor);
				}
			});
		};

		traverse(initialVertex);
		return visited;
	}

	depthFirstSearchIterative(initialVertex: string): Set<string> {
		if (!this.graph[initialVertex]) {
			throw new Error('Vertex not found in the graph');
		}

		const visited: Set<string> = new Set();
		const stack: string[] = [];

		stack.push(initialVertex);

		while (stack.length) {
			let vertex: string | undefined = stack.pop();

			if (!vertex) {
				break;
			}

			visited.add(vertex);

			this.graph[vertex].forEach(neighbor => {
				if (!visited.has(neighbor)) {
					stack.push(neighbor);
				}
			});
		}

		return visited;
	}

	breadthFirstSearch(initialVertex: string): Set<string> {
		if (!this.graph[initialVertex]) {
			throw new Error('Vertex not found in the graph');
		}

		const visited: Set<string> = new Set();
		const queue: string[] = [];

		queue.push(initialVertex);

		while (queue.length) {
			const vertex: string | undefined = queue.shift();

			if (!vertex) {
				break;
			}

			visited.add(vertex);

			this.graph[vertex].forEach(neighbor => {
				if (!visited.has(neighbor)) {
					queue.push(neighbor);
				}
			});
		}

		return visited;
	}
}

File: data-structures/graph/python/adjacency_list_graph.py
from typing import Dict, List


class AdjacencyListGraph:
    def __init__(self):
        self.graph: Dict[str, List[str]] = {}

    def add_vertex(self, vertex: str) -> None:
        self.graph[vertex] = []

    def add_edge(self, vertex1: str, vertex2: str) -> None:
        if vertex1 not in self.graph or vertex2 not in self.graph:
            raise ValueError('Vertex not found in the graph')
        self.graph[vertex1].append(vertex2)
        self.graph[vertex2].append(vertex1)

    def remove_vertex(self, vertex: str) -> None:
        if vertex not in self.graph:
            raise ValueError('Vertex not found in the graph')

        for vertex_list in self.graph.values():
            if vertex in vertex_list:
                vertex_list.remove(vertex)

        del self.graph[vertex]

    def remove_edge(self, vertex1: str, vertex2: str) -> None:
        if vertex1 not in self.graph or vertex2 not in self.graph:
            raise ValueError('Vertex not found in the graph')

        if vertex2 in self.graph[vertex1]:
            self.graph[vertex1].remove(vertex2)

        if vertex1 in self.graph[vertex2]:
            self.graph[vertex2].remove(vertex1)

    def depth_first_search_recursive(self, initial_vertex: str) -> List[str]:
        if initial_vertex not in self.graph:
            raise ValueError('Vertex not found in the graph')

        visited = []

        def traverse(vertex: str) -> None:
            visited.append(vertex)
            for neighbor in self.graph[vertex]:
                if neighbor not in visited:
                    traverse(neighbor)

        traverse(initial_vertex)
        return visited

    def depth_first_search_iterative(self, initial_vertex: str) -> List[str]:
        if initial_vertex not in self.graph:
            raise ValueError('Vertex not found in the graph')

        visited = []
        stack = [initial_vertex]

        while stack:
            vertex = stack.pop()
            if vertex not in visited:
                visited.append(vertex)
                stack.extend([n for n in self.graph[vertex] if n not in visited])

        return visited

    def breadth_first_search(self, initial_vertex: str) -> List[str]:
        if initial_vertex not in self.graph:
            raise ValueError('Vertex not found in the graph')

        visited = []
        queue = [initial_vertex]

        while queue:
            vertex = queue.pop(0)
            if vertex not in visited:
                visited.append(vertex)
                queue.extend([n for n in self.graph[vertex] if n not in visited])

        return visited
